### ISSUE TITLE
Getting '<head><script...' instead of data in response to XHR request 

### ISSUE THREAD

**Issue by Bohdan-Kalynovskyi** [cypress-io/cypress#24395]
### Current behavior

XHR response when running Cypress is: 

```
<head> \x3Cscript type='text/javascript'> document.domain = 'localhost'; !function(e){var t={};function n(r){if(t[r])return t[r].exports;var o=t[r]={i:r,l:!1,exports:{}};return e[r].call(o.exports,o,o.exports,n),o.l=!0,o.exports}n.m=e,n.c=t,n.d=function(e,t,r){n.o(e,t)||Object.defineProperty(e,t,{enumerable:!0,get:r})},n.r=function(e){"undefined"!=typeof Symbol&&Symbol.toStringTag&&Object.defineProperty(e,Symbol.toStringTag,{value:"Module"}),Object.defineProperty(e,"__esModule",{value:!0})},n.t=function(e,t){if(1&t&&(e=n(e)),8&t)return e;if(4&t&&"object"==typeof e&&e&&e.__esModule)return e;var r=Object.create(null);if(n.r(r),Object.defineProperty(r,"default",{enumerable:!0,value:e}),2&t&&"string"!=typeof e)for(var o in e)n.d(r,o,function(t){return e[t]}.bind(null,o));return r},n.n=function(e){var t=e&&e.__esModule?function(){return e.default}:function(){return e};return n.d(t,"a",t),t},n.o=function(e,t){return Object.prototype.hasOwnProperty.call(e,t)},n.p="",n(n.s=0)}({"./injection/main.js":function(e,t,n){"use strict";n.r(t);const r=window.Cypress=parent.Cypress;if(!r)throw new Error("Something went terribly wrong and we cannot proceed. We expected to find the global Cypress in the parent window but it is missing. This should never happen and likely is a bug. Please open an issue.");const o=(()=>{let e,t,n,r;const o=()=>{e=!1,t=!1,n=[],r={}},i=(e,t=[])=>"function"==typeof e?e.apply(window,t):window.eval(e);return o(),{wrap:()=>{const o={setTimeout:window.setTimeout,setInterval:window.setInterval,requestAnimationFrame:window.requestAnimationFrame,clearTimeout:window.clearTimeout,clearInterval:window.clearInterval},a=(e,t)=>o[e].apply(window,t),s=e=>n=>(t&&(r[n]=!0),a(e,[n])),u=t=>(...r)=>{let o,[s,u,...l]=r;return o=a(t,[(...r)=>{if(!e)return i(s,r);n.push({timerId:o,fnOrCode:s,params:r,type:t})},u,...l]),o};window.setTimeout=u("setTimeout"),window.setInterval=u("setInterval"),window.requestAnimationFrame=u("requestAnimationFrame"),window.clearTimeout=s("clearTimeout"),window.clearInterval=s("clearInterval")},reset:o,pause:a=>{e=Boolean(a),e||(t=!0,n.forEach(e=>{const{timerId:t,type:n,fnOrCode:o,params:a}=e;"setInterval"===n&&r[t]||i(o,a)}),o())}}})();r.removeAllListeners("app:timers:reset"),r.removeAllListeners("app:timers:pause"),r.on("app:timers:reset",o.reset),r.on("app:timers:pause",o.pause),o.wrap(),r.action("app:window:before:load",window)},0:function(e,t,n){e.exports=n("./injection/main.js")}}); \x3C/script> </head>{"domain":"wa2-localhost-dev-test.com","serverTime":1666736459210}
```

### Desired behavior

Expected response is:

```
{"domain":"wa2-localhost-dev-test.com","serverTime":1666736459210}
```

Obviously, html gets injected by Cypress. Request headers include 
```
 'Content-Type': 'application/json',
  Accept: 'application/json'
```
but unfortunately response headers include
`content-type: text/html; charset=utf-8`
I can't control the server unfortunately

### Test code to reproduce

Hope the bug is clear and I don't need to create a test repo 

### Cypress Version

10.11.0

### Node version

16.16.0

### Operating System

macOS 12.6.1

**Comment by Bohdan-Kalynovskyi** [comment]
This is a dupe of https://github.com/cypress-io/cypress/issues/2599. In that issue CSV response also had `content-type: text-html`

**Comment by Bohdan-Kalynovskyi** [comment]
Agree that normally json responses shouldn't have `content-type: text-html`, but what's the point of injecting into XHR responses?

**Comment by AtofStryker** [comment]
related to #24297 and should be fixed by #24414. I think we might need to also check the request content-type as well to make sure they are both html-like

**Comment by AtofStryker** [comment]
@Bohdan-Kalynovskyi are you able to share the request URL? I wonder if we need to do an additional check to make sure the request came from the Application Under Test frame which should fix the XHR injection problem

**Comment by AtofStryker** [comment]
@Bohdan-Kalynovskyi I don't believe we are going to be able to address this #24414 since the response type comes back with `text/html`, but I strongly feel that #24446 will solve this issue and we should be spiking into it in the next few weeks

**Comment by Bohdan-Kalynovskyi** [comment]
The request URL is http://localhost:3000/login/api/config/acceptance01 , it has the same domain as the main page.

Further thinking: I've seen weird techniques when js downloaded html file via XHR, and then did `document.write(stringContainingHTML)` (killing Cypress scripts inside the page which is not desirable, so I guess injecting into XHR body fixes exactly such issue). But the best way to handle such case is to inject during document.write, not XHR.

It would be also nice to have a flag that would disable such injections.

**Comment by AtofStryker** [comment]
>The request URL is http://localhost:3000/login/api/config/acceptance01 , it has the same domain as the main page. 

@Bohdan-Kalynovskyi thank you for sharing that. My guess is this is one of the initial requests that your application makes and cypress thinks it's the initial html load, so it tries to inject. Would you be able to make a quick reproduction repository that I can reproduce on my end to see if the fix I am proposing would solve the issue? 

> Further thinking: I've seen weird techniques when js downloaded html file via XHR, and then did document.write(stringContainingHTML) (killing Cypress scripts inside the page which is not desirable, so I guess injecting into XHR body fixes exactly such issue). But the best way to handle such case is to inject during document.write, not XHR.

> It would be also nice to have a flag that would disable such injections.

It's a bit more complicated than this unfortunately. The goal of this injection is to set the Cypress driver up in your application under test. The point is for this to happen on the initial `cy.visit`, but in this case the Cypress proxy things this XHR request is the initial visit 🙁. It's also the reason we can't have a flag for something like this, because it is core to Cypress working.

However I do feel good about the changes asked for in #24446 to resolve this issue!

**Comment by Bohdan-Kalynovskyi** [comment]
Couldn't create a simple repo to reproduce yet. 

Good news is that adding `Content-Type: application-json`. Fixed the issue for me, and hopefully, this thread would help others if similar things occur in future

**Comment by ranamowbey** [comment]
<head>
    <script type='text/javascript'>
        document.domain = '*****.com';
        (()=>{
            "use strict";
            const e = window.Cypress = parent.Cypress;
            if (!e)
                throw new Error("Something went terribly wrong and we cannot proceed. We expected to find the global Cypress in the parent window but it is missing. This should never happen and likely is a bug. Please open an issue.");
            const t = (()=>{
                let e, t, n, r;
                const i = ()=>{
                    e = !1,
                    t = !1,
                    n = [],
                    r = {}
                }
                  , o = (e,t=[])=>"function" == typeof e ? e.apply(window, t) : window.eval(e);
                return i(),
                {
                    wrap: ()=>{
                        const i = {
                            setTimeout: window.setTimeout,
                            setInterval: window.setInterval,
                            requestAnimationFrame: window.requestAnimationFrame,
                            clearTimeout: window.clearTimeout,
                            clearInterval: window.clearInterval
                        }
                          , a = (e,t)=>i[e].apply(window, t)
                          , s = e=>n=>(t && (r[n] = !0),
                        a(e, [n]))
                          , w = t=>(...r)=>{
                            let i, [s,w,...p] = r;
                            return i = a(t, [(...r)=>{
                                if (!e)
                                    return o(s, r);
                                n.push({
                                    timerId: i,
                                    fnOrCode: s,
                                    params: r,
                                    type: t
                                })
                            }
                            , w, ...p]),
                            i
                        }
                        ;
                        window.setTimeout = w("setTimeout"),
                        window.setInterval = w("setInterval"),
                        window.requestAnimationFrame = w("requestAnimationFrame"),
                        window.clearTimeout = s("clearTimeout"),
                        window.clearInterval = s("clearInterval")
                    }
                    ,
                    reset: i,
                    pause: a=>{
                        e = Boolean(a),
                        e || (t = !0,
                        n.forEach((e=>{
                            const {timerId: t, type: n, fnOrCode: i, params: a} = e;
                            "setInterval" === n && r[t] || o(i, a)
                        }
                        )),
                        i())
                    }
                }
            }
            )();
            e.removeAllListeners("app:timers:reset"),
            e.removeAllListeners("app:timers:pause"),
            e.on("app:timers:reset", t.reset),
            e.on("app:timers:pause", t.pause),
            t.wrap(),
            e.action("app:window:before:load", window)
        }
        )();
    </script>
</head>
ok



}

**Comment by ranamowbey** [comment]
i am getting this error , can someone give me a solution? this happens after i migrated from cypress 9.5 to 13

**Comment by AtofStryker** [comment]
@ranamowbey can you provide more details on your request that is having the code injected into it?
